File: stream-chat-android-docs/src/main/java/io/getstream/chat/docs/java/client/docusaurus/ListeningForEvents.java
package io.getstream.chat.docs.java.client.docusaurus;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import io.getstream.chat.android.client.ChatClient;
import io.getstream.chat.android.client.channel.ChannelClient;
import io.getstream.chat.android.client.events.ChatEvent;
import io.getstream.chat.android.client.events.NewMessageEvent;
import io.getstream.chat.android.client.events.NotificationRemovedFromChannelEvent;
import io.getstream.chat.android.models.Member;
import io.getstream.chat.android.models.Message;
import io.getstream.chat.android.client.utils.observable.Disposable;

/**
 * @see <a href="https://getstream.io/chat/docs/sdk/android/client/guides/listening-for-events/">Listening for Events</a>
 */
public class ListeningForEvents {

    class ListeningForEventsSnippet1 {

        public void listeningForEvents() {
            ChatClient chatClient = ChatClient.instance();
            ChannelClient channelClient = chatClient.channel("messaging", "general");

            // Subscribe for new message events
            Disposable disposable = channelClient.subscribeFor(
                    new Class[]{NewMessageEvent.class},
                    (ChatEvent event) -> {
                        Message message = ((NewMessageEvent) event).getMessage();
                    }
            );

            // Dispose when you want to stop receiving events
            disposable.dispose();
        }
    }

    class ListeningForEventsSnippet2 extends Fragment {

        @Override
        public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            // Setup chat screen components

            ChatClient chatClient = ChatClient.instance();
            ChannelClient channelClient = chatClient.channel("messaging", "general");
            channelClient.subscribeFor(
                    new Class[]{NotificationRemovedFromChannelEvent.class},
                    (ChatEvent event) -> {
                        Member member = ((NotificationRemovedFromChannelEvent) event).getMember();

                        String removedUserId = member.getUser().getId();
                        String currentUserId = chatClient.getCurrentUser().getId();
                        if (removedUserId.equals(currentUserId)) {
                            // Close the current chat screen as the current user has been removed
                            requireActivity().getOnBackPressedDispatcher().onBackPressed();
                        }
                    }
            );
        }
    }
}

File: stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/client/docusaurus/ListeningForEvents.kt
// ktlint-disable filename

package io.getstream.chat.docs.kotlin.client.docusaurus

import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import io.getstream.chat.android.client.ChatClient
import io.getstream.chat.android.client.channel.subscribeFor
import io.getstream.chat.android.client.events.NewMessageEvent
import io.getstream.chat.android.client.events.NotificationRemovedFromChannelEvent
import io.getstream.chat.android.client.utils.observable.Disposable

/**
 * [Listening for Events](https://getstream.io/chat/docs/sdk/android/client/guides/listening-for-events/)
 */
class ListeningForEventsSnippets {

    class ListeningForEventsSnippet1 {

        fun listeningForEvents() {
            val chatClient = ChatClient.instance()
            val channelClient = chatClient.channel("messaging", "general")

            // Subscribe for new message events
            val disposable: Disposable = channelClient.subscribeFor<NewMessageEvent> { newMessageEvent ->
                val message = newMessageEvent.message
            }

            // Dispose when you want to stop receiving events
            disposable.dispose()
        }
    }

    class ListeningForEventsSnippet2 : Fragment() {

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            // Setup chat screen components

            val chatClient = ChatClient.instance()
            val channelClient = chatClient.channel("messaging", "general")
            channelClient.subscribeFor<NotificationRemovedFromChannelEvent>(viewLifecycleOwner) { event ->
                val removedUserId = event.member.user.id
                val currentUserId = chatClient.getCurrentUser()?.id
                if (removedUserId == currentUserId) {
                    // Close the current chat screen as the current user has been removed
                    requireActivity().onBackPressedDispatcher.onBackPressed()
                }
            }
        }
    }
}
